src/DocumentDB.ChangeFeedProcessor/PartitionManagement/RemainingWorkEstimator.cs

```diff
@@ -19,69 +22,73 @@ internal class RemainingWorkEstimator : IRemainingWorkEstimator
-                IFeedResponse<Document> response = null;
+                return new List<RemainingPartitionWork>().AsReadOnly();
+            }
 
-                try
+            var tasks = Partitioner.Create(leases)
+                .GetPartitions(this.degreeOfParallelism)
+                .Select(partition => Task.Run(async () =>
                 {
-                    response = await query.ExecuteNextAsync<Document>().ConfigureAwait(false);
-                    long parsedLSNFromSessionToken = TryConvertToNumber(ExtractLSNFromSessionToken(response.SessionToken));
-                    long lastQueryLSN = response.Count > 0 ?
-                        TryConvertToNumber(GetFirstDocument(response).GetPropertyValue<string>(LSNPropertyName)) - 1
-                        : parsedLSNFromSessionToken;
-                    if (lastQueryLSN < 0)
+                    IList<RemainingPartitionWork> results = new List<RemainingPartitionWork>();
+                    using (partition)
                     {
-                        // Could not parse LSN from document, we cannot determine the amount of changes but since the query returned 1 document, we know it's at least 1
-                        remainingWork += 1;
-                        continue;
+                        while (partition.MoveNext())
+                        {
+                            ILease item = partition.Current;
+                            try
+                            {
+                                if (string.IsNullOrEmpty(item.PartitionId)) continue;
+                                var result = await this.GetRemainingWorkForLeaseAsync(item);
+                                results.Add(new RemainingPartitionWork(item.PartitionId, result));
+                            }
+                            catch (DocumentClientException ex)
+                            {
+                                Logger.WarnException($"Getting estimated work for {item.PartitionId} failed!", ex);
+                            }
+                        }
                     }
 
-                    long partitionRemainingWork = parsedLSNFromSessionToken - lastQueryLSN;
-                    remainingWork += partitionRemainingWork < 0 ? 0 : partitionRemainingWork;
-                }
-                catch (DocumentClientException clientException)
-                {
-                    Logger.WarnException("GetEstimateWork > exception: partition '{0}'", clientException, existingLease.PartitionId);
-                }
-            }
+                    return results.ToArray();
+                })).ToArray();
 
-            if (!hasAtLeastOneLease)
+            try
+            {
+                await Task.WhenAll(tasks);
+            }
+            catch (Exception ex)
             {
-                return 1;
+                Logger.WarnException("Incomplete estimation results!", ex);
```

Comment:
We should not be catching/eating all exceptions. It's strange that StyleCop/FxCop doesn't catch that. We already catch DocumentClientException inside each task, what else do we expect to fail here?

Reply:
there could be issues in runtime or in SDK -> so my point was just to log that there was the exception for some partition and rethrow. 

examples: throwing JsonSerialization exception when parsing error from bad request response (400) -> see the email I sent about returning HTML from the DocDB node

or a runtime configuration exception, etc... who knows ...